consensus/core/src/block_manager.rs

```diff
@@ -1,53 +1,397 @@
-    pub(crate) fn add_blocks(&mut self, mut blocks: Vec<VerifiedBlock>) -> Vec<VerifiedBlock> {
-        // TODO: add implementation - for now just dummy/test. Accept everything assuming history exists and cache them
-        // to ensure that they are not returned if they already processed.
-        blocks.sort_by_key(|b1| b1.round());
-        blocks
-            .into_iter()
-            .flat_map(|block| {
-                self.accepted_blocks
-                    .insert(block.reference())
-                    .then_some(block)
-            })
-            .collect()
+    pub(crate) fn accept_blocks(
+        &mut self,
+        mut blocks: Vec<VerifiedBlock>,
+    ) -> ConsensusResult<Vec<VerifiedBlock>> {
+        let mut accepted_blocks: Vec<VerifiedBlock> = vec![];
+
+        blocks.sort_by_key(|b| b.round());
+        for block in blocks {
+            if let Some(block) = self.try_accept_block(block)? {
+                // Try to unsuspend and accept any children blocks
+                let mut children_blocks = self.try_unsuspend_children_blocks(&block);
+                children_blocks.push(block);
+
+                // Accept the state in DAG here so the next block to be processed can find any DAG/store any accepted blocks.
+                self.dag_state
+                    .write()
+                    .accept_blocks(children_blocks.clone());
+
+                accepted_blocks.extend(children_blocks);
+            }
+        }
+
+        accepted_blocks.sort_by_key(|b| b.reference());
+        Ok(accepted_blocks)
+    }
+
+    /// Tries to accept the provided block. To accept a block its ancestors must have been already successfully accepted. If
+    /// block is accepted then Some result is returned. None is returned when either the block is suspended or the block
+    /// has been already accepted before.
+    fn try_accept_block(&mut self, block: VerifiedBlock) -> ConsensusResult<Option<VerifiedBlock>> {
+        let block_ref = block.reference();
+        let mut missing_ancestors = HashSet::new();
+
+        // If block has been already received and suspended, or already processed and stored, or is a genesis block, then skip it.
+        if self.suspended_blocks.contains_key(&block_ref)
+            || self
+                .dag_state
+                .read()
+                .contains_block_in_cache_or_store(&block_ref)?
+            || self.genesis.contains(&block_ref)
+        {
+            return Ok(None);
+        }
+
+        // make sure that we have all the required ancestors in store
+        for ancestor in block.ancestors() {
+            if self.genesis.contains(ancestor) {
+                continue;
+            }
+
+            // If the reference is already missing, or is not included in the store then we mark the block as non processed
+            // and we add the block_ref dependency to the `missing_blocks`
+            if self.missing_ancestors.contains_key(ancestor)
+                || !self
```

Comment:
I've considered checking first in the `self.missing_ancestors` as the least expensive check first , otherwise for blocks that are missing we'll need to touch the store/disk always. So ideally we would like to:
* check at bulk against the `self.missing_ancestors`
* for those which are not found there, check at bulk on the DagState
* merge the results
* now iterate

I've coded the above and I think it beared a little bit of complexity for not sure clear gains right now. So I've simplified and did as you said, just check directly in DagState. We can iterate on that later.